### PR TITLE
feat: show streak milestones modals when alert is on

### PR DESCRIPTION
**Pull request by kopancek**
## Changes

- Adds a new hook to show the reading streaks milestone modals

### Screenshots

Screenshot 2024-02-16 at 16 41 08
Screenshot 2024-02-16 at 16 40 40

## Events

Did you introduce any new tracking events?
Don't forget to update the [Analytics Taxonomy sheet](https://docs.google.com/spreadsheets/d/18Lv7zXges9QfVX5VYL1a-Hyl0e1sQ3sLr0OK8YZWKXI/edit#gid=0)

Log the new/changed events below:

| Type   | event_name  | value |
|--------|-------------|-------|
| Change/New | event name  | extra: { ... } |

### On those affected packages:
- [X] Have you done sanity checks in the webapp?
- [ ] Have you done sanity checks in the extension?
- [ ] Does this not break anything in companion?

### Did you test the modified components media queries?
- [X] MobileL (420px)
- [X] Tablet (656px)
- [X] Laptop (1020px)

#### Did you test on actual mobile devices?
- [ ] iOS (Chrome and Safari)
- [ ] Android

MI-75 #done

### CODE REVIEW

File: packages/shared/src/hooks/streaks/useStreakMilestone.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+  useEffect(() => {
+    if (
+      !loadedAlerts ||
+      alerts?.showStreakMilestone !== true ||
+      isLoading ||
+      !isEnabled ||
+      !loadedSettings ||
+      optOutWeeklyGoal ||
+      !streak?.current
```
Comment (rebelchris) [inline packages/shared/src/hooks/streaks/useStreakMilestone.ts:27]: Are the ors really what we want?
Feels like some are or and some should be and no?

It's a bit hard to follow due to the big evaluation.

@capJavert Possibly would like your review on this hook as well :)

Comment (rebelchris) [reply packages/shared/src/hooks/streaks/useStreakMilestone.ts:27]: Alternative was thinking if we could simply put the high level exits (as optOut above the useEffect even?

File: packages/shared/src/hooks/streaks/useStreakMilestone.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    openModal({
+      type: modalType,
+      props: {
+        currentStreak: streak?.current,
+        onRequestClose: () => {
+          updateAlerts({ showStreakMilestone: false });
+          closeModal();
+        },
+      } as StreakModalProps,
```
Comment (rebelchris) [inline packages/shared/src/hooks/streaks/useStreakMilestone.ts:40]: Because the dynamic modalType Ts issue?

File: packages/shared/src/hooks/useIncrementReadingRank.ts [hunk 3]
```diff
@@ -16,14 +17,27 @@ export default function useIncrementReadingRank(): ReturnType {
   const { user } = useContext(AuthContext);
   const queryKeyRef = useRef<string[]>();
   const queryClient = useQueryClient();
-  const [clearQueries] = useDebounce(() => {
-    queryClient.invalidateQueries(queryKeyRef.current);
-  }, 100);
+  const [clearQueries] = useDebounce(
+    async (readToday = false): Promise<void> => {
+      const promises = [queryClient.invalidateQueries(queryKeyRef.current)];
```
Comment (rebelchris) [inline packages/shared/src/hooks/useIncrementReadingRank.ts:22]: Is there any change this could result in accidentally clearing all queries (eg if it evaluates null?) (not sure just asking to be safe)

File: packages/shared/src/hooks/useIncrementReadingRank.ts [hunk 4]
```diff
@@ -16,14 +17,27 @@ export default function useIncrementReadingRank(): ReturnType {
...
+      if (!readToday) {
+        promises.push(
+          queryClient.invalidateQueries(
+            generateQueryKey(RequestKey.UserStreak, user),
+          ),
+        );
+      }
+      await Promise.all(promises);
+    },
+    100,
```
Comment (rebelchris) [inline packages/shared/src/hooks/useIncrementReadingRank.ts:32]: Why do we debounce it actually? (product req or engineering thing)

File: packages/shared/src/lib/featureManagement.ts [hunk 5]
```diff
@@ -24,10 +24,7 @@ const feature = {
   lowImps: new Feature('feed_low_imps'),
   bookmarkOnCard: new Feature('bookmark_on_card', false),
   feedLayout: new Feature('feed_layout', FeedLayout.Control),
-  readingStreaks: new Feature(
-    'reading_streaks',
-    ReadingStreaksExperiment.Control,
-  ),
+  readingStreaks: new Feature('reading_streaks', ReadingStreaksExperiment.V1),
```
Comment (rebelchris) [inline packages/shared/src/lib/featureManagement.ts:27]: Just be careful to not push this please :)

File: packages/shared/src/hooks/streaks/useStreakMilestone.ts (hunk 1 above)
Comment (capJavert) [reply packages/shared/src/hooks/streaks/useStreakMilestone.ts:27]: yeah, basically two things:

1. Minimize amount of dependencies by precomputing conditions outside of hook. Eg. instead of doing:
```js
useEffect(() => {
  if (cond1 && cond2 && !cond3 && cond4) {
    return
  }

  // logic
}, [cond1, cond2, cond3, cond4]);
```

do this:
```js
const condition = cond1 && cond2 && !cond3 && cond4

useEffect(() => {
  if (!condition) {
    return
  }

  // logic
}, [condition]);
```
`condition` variable name should reflect what the condition means. Also it lowers amount of dependencies that React has to check thus lowering number of `useEffect` executions.

2. As Chris said, the condition sometimes might be complicated for single evaluation and few early returns make it more readable so then it would look something like:
```js
const condition1 = cond1 && cond2

const condition2 = !cond3 && cond4

useEffect(() => {
  if (!condition1) {
    return
  }

  if (!condition2) {
    return
  }

  // logic
}, [condition1, condition2]);
```

Point 2. can be optional here but you can probably decide better since you are more in line on what code needs to do 👍

File: packages/shared/src/hooks/streaks/useStreakMilestone.ts (hunk 2 above)
Comment (capJavert) [reply packages/shared/src/hooks/streaks/useStreakMilestone.ts:40]: We have a generic available on openModal right, or does that also not work?

File: packages/shared/src/hooks/useIncrementReadingRank.ts (hunk 3 above)
Comment (capJavert) [reply packages/shared/src/hooks/useIncrementReadingRank.ts:22]: Yep, if `queryKeyRef.current` is `[]` it will invalidate all queries, we should probably check if empty to avoid it.

**Review by nensidosari — COMMENTED** [review]
Nth to add except what's already being discussed. 🙌

File: packages/shared/src/hooks/streaks/useStreakMilestone.ts (hunk 2 above)
Comment (sshanzel) [reply packages/shared/src/hooks/streaks/useStreakMilestone.ts:40]: This is probably because the modal has not yet been introduced to the list of modals at `/components/modals/common.tsx`

File: packages/shared/src/components/Feed.tsx [hunk 6]
```diff
@@ -263,6 +264,8 @@ export default function Feed<T>({
     }
   }, [selectedPost]);
 
+  useStreakMilestone();
```
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/shared/src/components/Feed.tsx:267]: Should probably be at MainLayout? This is so when the user is visiting a post page, and clicked read article, it would count as view - the milestone popup would be possible to be triggered.

**Review by sshanzel — COMMENTED** [review]
Looking great overall! 🥳

File: packages/shared/src/hooks/streaks/useStreakMilestone.ts (hunk 1 above)
Comment (kopancek) [reply packages/shared/src/hooks/streaks/useStreakMilestone.ts:27]: We cannot put opt outs above hooks, because all hoooks need to be evaluated, they cannot be conditionally skipped: https://legacy.reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level

File: packages/shared/src/components/Feed.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (kopancek) [reply packages/shared/src/components/Feed.tsx:267]: To show the modals, the hook needs to be actually executed, which is done on render. If we put it on `MainLayout`, how do we trigger the render? AFAIK we rarely render the whole layout.

File: packages/shared/src/hooks/streaks/useStreakMilestone.ts (hunk 2 above)
Comment (kopancek) [reply packages/shared/src/hooks/streaks/useStreakMilestone.ts:40]: Both of the modal types were introduced in `components/modals/common.tsx`. This is the error I was getting from TS compiler:
```
Object literal may only specify known properties, and 'onRequestClose' does not exist in type 'LazyModalComponentType<LazyModal.SquadMember> | LazyModalComponentType<LazyModal.SquadTour> | LazyModalComponentType<LazyModal.UpvotedPopup> | ... 13 more ... | LazyModalComponentType<...>'.ts(2353)
common.tsx(161, 9): The expected type comes from property 'props' which is declared here on type 'LazyModalType<LazyModal>'
```

File: packages/shared/src/hooks/useIncrementReadingRank.ts (hunk 3 above)
Comment (kopancek) [reply packages/shared/src/hooks/useIncrementReadingRank.ts:22]: There's no change in behavior here from the existing one, so this risk was already there. Do we want to fix it?

File: packages/shared/src/hooks/useIncrementReadingRank.ts (hunk 4 above)
Comment (kopancek) [reply packages/shared/src/hooks/useIncrementReadingRank.ts:32]: I didn't dig deeper here, I just replicated the existing behavior. Will try to understand it now...

File: packages/shared/src/hooks/streaks/useStreakMilestone.ts (hunk 1 above)
Comment (kopancek) [reply packages/shared/src/hooks/streaks/useStreakMilestone.ts]: Please take a look again. I removed unneeded dependencies in the condition, split it into a separate variable and rewritten the code into multiple lines with comments.

File: packages/shared/src/hooks/useIncrementReadingRank.ts (hunk 4 above)
Comment (kopancek) [reply packages/shared/src/hooks/useIncrementReadingRank.ts:32]: The setTimeout and clearing of queries was introduced here: https://github.com/dailydotdev/apps/commit/630e6ca4d5799be56d87067314b787a2e09b2857#diff-2ca4b5627b297d8adac0ec5fa73626249c395ceb53225a067d128ed3fad2a6d5

Maybe it was because of animations? 🤔

File: packages/shared/src/hooks/useIncrementReadingRank.ts (hunk 3 above)
Comment (kopancek) [reply packages/shared/src/hooks/useIncrementReadingRank.ts:22]: Looking at the code again, `queryKeyRef.current` should never be empty, as we assign it before calling the clearQueries function.

```typescript
const queryKey = getRankQueryKey(user);
queryKeyRef.current = queryKey;	      queryKeyRef.current = queryKey;
```

File: packages/shared/src/components/Feed.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (kopancek) [reply packages/shared/src/components/Feed.tsx:267]: Asking product if it's OK to show these modals on feed pages only: https://dailydotdev.slack.com/archives/C02E2C3C13R/p1708421020306569?thread_ts=1708099375.192589&cid=C02E2C3C13R

[the remark this answers is not in the document]
Comment (rebelchris) [reply packages/shared/src/components/Feed.tsx:267]: Not sure I follow your comment, we have many of such hooks in the past on mainLayout afaik :) 
(the modals that show on action)

@sshanzel will know a lot about this.
Don't see why we would have to limit ourselves.

File: packages/shared/src/hooks/useIncrementReadingRank.ts (hunk 3 above)
Comment (rebelchris) [reply packages/shared/src/hooks/useIncrementReadingRank.ts:22]: Perhaps potentially if user is null?
Not sure but if it's easy to simply check for null value it would be appreciated.

File: packages/shared/src/hooks/useIncrementReadingRank.ts (hunk 4 above)
Comment (rebelchris) [reply packages/shared/src/hooks/useIncrementReadingRank.ts:32]: Hmm probably a edge-case before feel free to remove for your needs if we can get away with not having it for the new one.
But guess we support both so then let's leave to avoid it (can't fully recall)

File: packages/shared/src/hooks/streaks/useStreakMilestone.ts (hunk 2 above)
Comment (rebelchris) [reply packages/shared/src/hooks/streaks/useStreakMilestone.ts:44]: It's because we omit them.
```
type LazyModalComponentType<K extends LazyModal> = Omit<
  GetComponentProps<ModalsType[K]>,
  'isOpen' | 'onRequestClose'
>;
```

I think the idea was that these should never be interfered with as they do a lot of magic stuff.
Some alternatives is to pass a onAfterClose *(some modal has this already)

File: packages/shared/src/hooks/streaks/useStreakMilestone.ts (hunk 1 above)
Comment (capJavert) [reply packages/shared/src/hooks/streaks/useStreakMilestone.ts]: From my point of view good idea with `||=`. I would only check if we polyfill it correctly. The adoption is good but since some browsers got the support pretty recently https://caniuse.com/?search=Logical%20OR%20assignment I would want to be sure. Maybe check in compiled bundle how it looks?

Comment (kopancek) [reply packages/shared/src/hooks/streaks/useStreakMilestone.ts]: Oh, right. TIL that browsers didn't support that assignment before 😊 

I think I'll change to assigning directly. It's just syntactic sugar...

File: packages/shared/src/hooks/streaks/useStreakMilestone.ts (hunk 2 above)
Comment (kopancek) [reply packages/shared/src/hooks/streaks/useStreakMilestone.ts]: TIL, thanks for the input 👍

File: packages/shared/src/components/Feed.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/shared/src/components/Feed.tsx:267]: Yeah, also, there can be a case when the user visits a post page link -> clicks read article -> reads the article in a different tab -> comes back to the post page tab -> popup should show.

> AFAIK we rarely render the whole layout.

The `MainLayout` is rendered in almost every page, probably 96% of all pages (sample excluded ones are `onboarding`, referral page, etc.).

[the remark this answers is not in the document]
Comment (kopancek) [reply packages/shared/src/components/Feed.tsx]: What I meant by that comment is that when you click on the read post button, the main layout is not re-rendered, so the hook to show the modals would not be executed, increasing the delay that we show the modals.

I have moved the hook back to the `MainLayout` though, please take another look 🙏

[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/shared/src/components/Feed.tsx]: It would be re-rendered when you go back to the app (window focus) since the boot query will refetch, and it is inside the boot provider which will trigger the re-render and MainLayout is wrapped by the BootProvider.

File: packages/shared/src/components/MainLayout.tsx [hunk 7]
```diff
@@ -87,6 +88,7 @@ function MainLayoutComponent({
   useAuthVerificationRecovery();
   useNotificationParams();
   useReferralReminder();
+  useStreakMilestone();
```
Comment (rebelchris) [inline packages/shared/src/components/MainLayout.tsx:91]: Nice!!

File: packages/shared/src/hooks/streaks/useStreakMilestone.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  useEffect(() => {
+    if (shouldHideModal) {
+      return;
+    }
+
+    openModal({
+      type: modalType,
+      props: {
+        currentStreak: streak?.current,
+        onAfterClose: () => {
```
Comment (rebelchris) [inline packages/shared/src/hooks/streaks/useStreakMilestone.ts:39]: Was this a default one we had? Else im missing the prop. But might have missed this nuance myself 😂

**Review by rebelchris — APPROVED** [review]
Nothing blocking, basic question

File: packages/shared/src/hooks/streaks/useStreakMilestone.ts (hunk 8 above)
Comment (kopancek) [reply packages/shared/src/hooks/streaks/useStreakMilestone.ts:39]: It's defined on `ReactModal.Props` that all our modals extend. So there's more that we probably could be using and didn't know about 😊

Comment (rebelchris) [reply packages/shared/src/hooks/streaks/useStreakMilestone.ts:39]: Good to know indeed!
Something storybook could enhance 🥲
